src/game/logic/object/object.cpp

```diff
@@ -2021,3 +2346,541 @@ void Object::Notify_Firing_Tracker_Shot_Fired(const Weapon *weapon, ObjectID id)
+
+        if (power != nullptr) {
+            power->Pause_Countdown(pause);
+        }
+    }
+}
+
+void Object::On_Disabled_Edge(bool edge)
+{
+    for (BehaviorModule **m = m_allModules; *m != nullptr; m++) {
+        (*m)->On_Disabled_Edge(edge);
+    }
+
+    DozerAIInterface *dozer;
+
+    if (Get_AI_Update_Interface() != nullptr) {
+        dozer = Get_AI_Update_Interface()->Get_Dozer_AI_Interface();
+    } else {
+        dozer = nullptr;
+    }
+
+    if (edge && dozer != nullptr && dozer->Get_Current_Task() != DOZER_TASK_INVALID) {
+        dozer->Cancel_Task(dozer->Get_Current_Task());
+    }
+
+    Player *player = Get_Controlling_Player();
+
+    if (player != nullptr) {
+        static NameKeyType radar = g_theNameKeyGenerator->Name_To_Key("RadarUpgrade");
+        RadarUpgrade *upgrade = static_cast<RadarUpgrade *>(Find_Module(radar));
+
+        if (upgrade != nullptr && upgrade->Is_Already_Upgraded()) {
+            bool proof = upgrade->Get_Disable_Proof();
+
+            if (edge) {
+                player->Remove_Radar(proof);
+            } else {
+                player->Add_Radar(proof);
+            }
+        }
+    }
+
+    int energy = Get_Template()->Get_Energy_Production();
+
+    if (energy > 0) {
+        static NameKeyType powerPlant = g_theNameKeyGenerator->Name_To_Key("PowerPlantUpgrade");
+        static NameKeyType overCharge = g_theNameKeyGenerator->Name_To_Key("OverchargeBehavior");
+        PowerPlantUpgrade *power = static_cast<PowerPlantUpgrade *>(Find_Module(powerPlant));
+
+        if (power != nullptr && power->Is_Already_Upgraded()) {
+            energy += Get_Template()->Get_Energy_Bonus();
+        }
+
+        OverchargeBehavior *charge = static_cast<OverchargeBehavior *>(Find_Module(overCharge));
+
+        if (charge != nullptr && charge->Is_Overcharge_Active()) {
+            energy += Get_Template()->Get_Energy_Bonus();
+        }
+
+        if (player != nullptr) {
+            player->Get_Energy()->Adjust_Power(energy, !edge);
+        }
+    }
+}
+
+void Object::Check_Disabled_Status()
+{
+    unsigned int frame = g_theGameLogic->Get_Frame();
+
+    for (int i = 0; i < DISABLED_TYPE_COUNT; i++) {
+        if (Get_Disabled_State(static_cast<DisabledType>(i)) && frame >= m_disabledStateFrames[i]) {
+            Clear_Disabled(static_cast<DisabledType>(i));
+            m_disabledStates.Set(i, false);
+        }
+    }
+}
+
+void Object::Attempt_Healing(float f, const Object *obj)
+{
+    BodyModuleInterface *body = Get_Body_Module();
+
+    if (body != nullptr) {
+        DamageInfo info;
+        info.m_in.m_damageType = DAMAGE_HEALING;
+        info.m_in.m_deathType = DEATH_NONE;
+        ObjectID id;
+
+        if (obj) {
+            id = obj->Get_ID();
+        } else {
+            id = OBJECT_UNK;
+        }
+
+        info.m_in.m_sourceID = id;
+        info.m_in.m_amount = f;
+        body->Attempt_Healing(&info);
+    }
+}
+
+void Object::On_Removed_From(Object *removed)
```

Comment:
`removed` is unused. Is function body correct?

Reply:
Yes.